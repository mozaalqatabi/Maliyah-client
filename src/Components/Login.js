import React, { useEffect, useState } from 'react';
import { Card, Form, Button, Alert, Spinner } from 'react-bootstrap';  // React-Bootstrap components
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { Link, useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getUser, resetFlags } from '../Features/UserSlice';
import { LoginValidation } from '../Validations/LoginValidation';

function Login() {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const message = useSelector((state) => state.counter.message);
  const user = useSelector((state) => state.counter.user);
  const isSuccess = useSelector((state) => state.counter.isSuccess);
  const isError = useSelector((state) => state.counter.isError);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const {
    register,
    handleSubmit: submitForm,
    formState: { errors },
  } = useForm({ resolver: yupResolver(LoginValidation) });

  const [isSubmitting, setIsSubmitting] = useState(false);  // Track form submission state

  const handleSubmit = () => {
    const user = {email: email.toLowerCase(), password };
    setIsSubmitting(true);
    dispatch(resetFlags());
    dispatch(getUser(user));
  };

  useEffect(() => {
  if (isSuccess && user && Object.keys(user).length > 0) {
    localStorage.setItem("user", JSON.stringify(user));

    const tutorialKey = `tutorialShown_${user.email}`;
    const justRegistered = localStorage.getItem("justRegistered");

    // ✅ Clean up registration flag
    if (justRegistered === "true") {
      localStorage.removeItem("justRegistered");
    }

    // ✅ First-time user or fresh registration
    if (justRegistered === "true" || !localStorage.getItem(tutorialKey)) {
      navigate("/dashboard", { state: { showTutorial: true } });
    } else {
      navigate("/dashboard");
    }

    dispatch(resetFlags());
    setIsSubmitting(false);
  } else if (isError) {
    setIsSubmitting(false);
  }
}, [isSuccess, user, isError, navigate, dispatch]);


  return (
    <div
      style={{
        backgroundColor: "#A6BDFF",
        minHeight: "100vh",
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
      }}
    >
      <Link
        to="/"
        style={{
          position: "absolute",
          top: "50px",
          left: "100px",
          color: "#4F6BED",
          textDecoration: "none",
          fontWeight: "500",
          fontSize: "0.95rem",
        }}
      >
        ← Back to Home Page
      </Link>

      <div
        style={{
          width: "100%",
          maxWidth: "420px",
          backgroundColor: "white",
          padding: "30px",
          borderRadius: "12px",
          boxShadow: "0 0 20px rgba(0,0,0,0.1)",
        }}
      >
        <div className="text-center mb-4">
          <h4 className="mt-3 mb-1">Sign in to Maliyah</h4>
          <p style={{ fontSize: "0.9rem", color: "#666" }}>
            Enter your details to access your account
          </p>
        </div>

        <form onSubmit={submitForm(handleSubmit)}>
          {/* Email Input */}
          <Form.Group controlId="email" className="mb-3">
            <Form.Label>Email</Form.Label>
            <Form.Control
              type="email"
              placeholder="Enter your email"
              {...register("email", {
                value: email,
                onChange: (e) => setEmail(e.target.value),
              })}
              isInvalid={!!errors.email}
              disabled={isSubmitting}
            />
            <Form.Control.Feedback type="invalid">
              {errors.email?.message}
            </Form.Control.Feedback>
          </Form.Group>

          {/* Password Input */}
          <Form.Group controlId="password" className="mb-3">
            <Form.Label>Password</Form.Label>
            <Form.Control
              type="password"
              placeholder="Enter your password"
              {...register("password", {
                value: password,
                onChange: (e) => setPassword(e.target.value),
              })}
              isInvalid={!!errors.password}
              disabled={isSubmitting}
            />
            <Form.Control.Feedback type="invalid">
              {errors.password?.message}
            </Form.Control.Feedback>
          </Form.Group>

          {/* Submit Button */}
          <Button
            color="primary"
            className="form-control mb-3"
            type="submit"
            style={{ backgroundColor: "#4F6BED", border: "none" }}
            disabled={isSubmitting}
          >
            {isSubmitting ? (
              <Spinner
                as="span"
                animation="border"
                size="sm"
                role="status"
                aria-hidden="true"
              />
            ) : (
              "Sign In"
            )}
          </Button>

          {/* Error Message */}
          {isError && (
            <Alert variant="danger" className="mt-3 text-center">
              {message || "Invalid email or password"}
            </Alert>
          )}

          {/* Links */}
          <div className="text-center mt-3">
            <span className="text-muted">
              <Link to="/forgot-password" className="text-decoration-none">
                Forgot Password?
              </Link>
            </span>
          </div>

          <div className="text-center mt-3">
            <small>
              Don't have an account?{" "}
              <Link to="/register" style={{ color: "#4F6BED" }}>
                Sign Up Now
              </Link>
            </small>
            <span className="text-muted"></span>
          </div>
        </form>
      </div>
    </div>
  );
}

export default Login;
