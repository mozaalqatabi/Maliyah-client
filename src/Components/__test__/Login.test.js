import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import Login from '../Login';

const mockStore = configureStore([]);
const store = mockStore({
  counter: { isSuccess: false, isError: false, user: null, message: '' },
});

beforeAll(() => {
  jest.spyOn(console, "warn").mockImplementation(() => {});
});

describe('Login Component Tests', () => {

  test('renders email and password fields', () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    // Check for Email field
    const emailInput = screen.getByLabelText(/email/i);
    expect(emailInput).toBeInTheDocument();
    expect(emailInput).toHaveAttribute('type', 'email');

    // Check for Password field
    const passwordInput = screen.getByLabelText(/password/i);
    expect(passwordInput).toBeInTheDocument();
    expect(passwordInput).toHaveAttribute('type', 'password');
  });

  test('allows typing in email and password fields', () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const emailInput = screen.getByLabelText(/email/i);
    const passwordInput = screen.getByLabelText(/password/i);

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });

    expect(emailInput.value).toBe('test@example.com');
    expect(passwordInput.value).toBe('password123');
  });

  test('validates email format using regex', () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const emailInput = screen.getByLabelText(/email/i);
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

    // Valid email format
    fireEvent.change(emailInput, { target: { value: 'valid.email@example.com' } });
    expect(emailRegex.test(emailInput.value)).toBe(true);

    // Invalid email format
    fireEvent.change(emailInput, { target: { value: 'invalid-email' } });
    expect(emailRegex.test(emailInput.value)).toBe(false);
  });

  test('shows validation errors for empty fields on submit', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    const submitButton = screen.getByRole('button', { name: /sign in/i });
    fireEvent.click(submitButton);

    // Wait for validation messages
    expect(await screen.findByText(/email should not be empty/i)).toBeInTheDocument();
    expect(await screen.findByText(/password should not be empty/i)).toBeInTheDocument();
  });

  test('matches the UI snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    // Snapshot testing
    expect(container).toMatchSnapshot();
  });

});



