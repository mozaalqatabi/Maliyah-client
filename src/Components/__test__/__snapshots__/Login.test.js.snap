// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component Tests matches the UI snapshot 1`] = `
<div>
  <div
    style="background-color: rgb(166, 189, 255); min-height: 100vh; display: flex; justify-content: center; align-items: center;"
  >
    <a
      href="/"
      style="position: absolute; top: 50px; left: 100px; color: rgb(79, 107, 237); text-decoration: none; font-weight: 500; font-size: 0.95rem;"
    >
      ‚Üê Back to Home Page
    </a>
    <div
      style="width: 100%; max-width: 420px; background-color: white; padding: 30px; border-radius: 12px; box-shadow: 0 0 20px rgba(0,0,0,0.1);"
    >
      <div
        class="text-center mb-4"
      >
        <h4
          class="mt-3 mb-1"
        >
          Sign in to Maliyah
        </h4>
        <p
          style="font-size: 0.9rem; color: rgb(102, 102, 102);"
        >
          Enter your details to access your account
        </p>
      </div>
      <form>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="email"
          >
            Email
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            placeholder="Enter your email"
            type="email"
          />
          <div
            class="invalid-feedback"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder="Enter your password"
            type="password"
          />
          <div
            class="invalid-feedback"
          />
        </div>
        <button
          class="form-control mb-3 btn btn-primary"
          color="primary"
          style="background-color: rgb(79, 107, 237);"
          type="submit"
        >
          Sign In
        </button>
        <div
          class="text-center mt-3"
        >
          <span
            class="text-muted"
          >
            <a
              class="text-decoration-none"
              href="/forgot-password"
            >
              Forgot Password?
            </a>
          </span>
        </div>
        <div
          class="text-center mt-3"
        >
          <small>
            Don't have an account?
             
            <a
              href="/register"
              style="color: rgb(79, 107, 237);"
            >
              Sign Up Now
            </a>
          </small>
          <span
            class="text-muted"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
